PluginDetails: Make plugin details page look good in topnav

torkelo opened the pull request:
* [x] Adds new headerExtra hook (bad name?) to NavModelItem to add a custom element to header after heading and subTitle 
* [x] Makes some minor design tweaks to PluginDetailsHeader to make the styles fit the page header better (normal font size & weight like description on other pages that are in similar location) 
* [x] Removes internal tab padding so that the content of each tab is spaced like other pages
* [x] Removes internal tab scroll container, whole page should scroll like other pages 

Just a quick first take on it could need some help refining this. 

Screenshots (with old nav)
![Screenshot from 2022-09-21 15-40-57](https://user-images.githubusercontent.com/10999/191520462-841fa950-7d47-45ca-9039-6b26b3f0f6c7.png)

Screenshots (with topnav)
![Screenshot from 2022-09-21 15-41-04](https://user-images.githubusercontent.com/10999/191520430-96398507-1f74-4d7c-9b45-eedb02ab85b4.png)

torkelo committed:
PluginDetails: Make plugin details page look good in topnav

torkelo commented on public/app/features/plugins/admin/pages/PluginDetails.tsx:
```diff
@@ -72,29 +72,24 @@ export default function PluginDetails({ match, queryParams }: Props): JSX.Elemen
 
-  return (
-    <Page>
+  const pageNav: NavModelItem = {
+    text: plugin.name,
+    img: plugin.info.logos.small,
+    breadcrumbs: [{ title: 'Plugins', url: parentUrl }],
+    children: tabs.map((tab) => ({
+      text: tab.label,
+      icon: toIconName(tab.icon ?? ''),
+      url: tab.href,
+      active: tab.id === pageId,
+    })),
```
I think this is just quick example of what the model should look like, I think what we should do is change usePluginDetailsTabs so it returns NavModelItem (so this is built inside a useMemo) 
Would like help with this if possible

jackw replied:
@torkelo - took a stab at this and refactored the pluginDetails hook together with @mckn. I believe it now does what we need for this PR. Let me know if I've missed anything.

torkelo replied:
@jackw that's great, thanks!

torkelo committed:
Minor style tweak aligning things

ashharrison90 reviewed (approved):
behaviour-wise i think this looks great! 🙌 

agreed about generating `pageNav` in the hook

mckn committed:
minor refactoring where I moved the logic to decide the default tab into its own hook.

jackw committed:
refactor(plugindetails): first pass at using navmodel for usePluginDetailsTabs hook

jackw committed:
refactor(plugindetails): move "reset page when uninstalling plugin" to installcontrols

this prevents a user from seeing a blank page if they uninstall an app plugin whilst viewing a
config page

mckn commented on public/app/features/plugins/admin/hooks/usePluginDetailsTabs.tsx:
```diff
@@ -1,104 +1,144 @@
+        text: PluginTabLabels.VERSIONS,
         id: PluginTabIds.VERSIONS,
-        href: `${pathname}?page=${PluginTabIds.VERSIONS}`,
+        icon: toIconName('history'),
```
I think we can remove the `toIconName` and use the valid IconName here directly.

jackw replied:
Ah thanks for spotting that. Have updated it. 👍

mckn commented on public/app/features/plugins/admin/hooks/usePluginDetailsTabs.tsx:
```diff
@@ -1,104 +1,144 @@
       });
     }
 
     if (canConfigurePlugins) {
```
nit: what about doing an early exit here? basically change this to something like:

```ts
if (!canConfigurePlugins) {
   return navModelChildren;
}
```

To prevent nesting.

jackw replied:
Nice suggestion, have updated it. 👍

jackw committed:
refactor(plugindetails): remove usage of toIconName and reduce nested if

jackw commented on packages/grafana-data/src/types/navModel.ts:
```diff
@@ -36,6 +36,8 @@ export interface NavModelItem extends NavLinkDTO {
   tabSuffix?: ComponentType<{ className?: string }>;
   hideFromBreadcrumbs?: boolean;
+  /** To render custom things between title and child tabs */
+  headerExtra?: ComponentType;
```
It's a tricky one! `headerCustomContent` or `headerAdditionalContent`? 🤷

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into plugin-details-page-topnav

torkelo committed:
Trying to fix tests

torkelo commented:
@mckn @jackw The tests are failing I fixed the ones that related to a change in how the tabs are rendered but these are also failing 

```ts
describe('viewed as user without grafana admin permissions', () => {
    beforeAll(() => {
      mockUserPermissions({
        isAdmin: false,
        isDataSourceEditor: false,
        isOrgAdmin: false,
      });
    });

    it("should not display an install button for a plugin that isn't installed", async () => {
      const { queryByRole, findByRole } = renderPluginDetails({ id, isInstalled: false });

      expect(await findByRole('tab', { name: `Tab ${PluginTabLabels.OVERVIEW}` })).toBeInTheDocument();

      expect(queryByRole('button', { name: /^install/i })).not.toBeInTheDocument();
    });
```

They are failing because the mocking of permissions is not having any effect InstallControls.tsx is using contextSrv.hasAccess and that is mocked here 
https://github.com/grafana/grafana/blob/main/public/app/features/plugins/admin/components/InstallControls/InstallControls.tsx#L26 to always return true.

torkelo committed:
minor fix

jackw commented:
> @mckn @jackw The tests are failing I fixed the ones that related to a change in how the tabs are rendered but these are also failing
> 
> ```ts
> describe('viewed as user without grafana admin permissions', () => {
>     beforeAll(() => {
>       mockUserPermissions({
>         isAdmin: false,
>         isDataSourceEditor: false,
>         isOrgAdmin: false,
>       });
>     });
> 
>     it("should not display an install button for a plugin that isn't installed", async () => {
>       const { queryByRole, findByRole } = renderPluginDetails({ id, isInstalled: false });
> 
>       expect(await findByRole('tab', { name: `Tab ${PluginTabLabels.OVERVIEW}` })).toBeInTheDocument();
> 
>       expect(queryByRole('button', { name: /^install/i })).not.toBeInTheDocument();
>     });
> ```
> 
> They are failing because the mocking of permissions is not having any effect InstallControls.tsx is using contextSrv.hasAccess and that is mocked here https://github.com/grafana/grafana/blob/main/public/app/features/plugins/admin/pages/PluginDetails.test.tsx#L55 to always return true.

Thanks for the ping - I will take a look shortly @torkelo !

mckn commented:
> 

Let me know if you need any help with this @jackw

jackw committed:
test(plugindetails): update selectors causing failing tests

jackw committed:
chore(plugindetails): remove commented out test code

jackw committed:
test(plugindetails): clean up - remove unnecesary usage of waitFor
